src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3524,22 +3524,36 @@ protected static boolean isSingleByteSearch(DynamicObject source, DynamicObject
     }
 
     @Primitive(name = "string_character_byte_index", needsSelf = false, lowerFixnum = 2)
-    @ImportStatic(StringGuards.class)
+    @ImportStatic({ StringGuards.class, StringOperations.class })
     public static abstract class CharacterByteIndexNode extends PrimitiveArrayArgumentsNode {
 
         @Specialization(guards = "isSingleByteOptimizable(string)")
-        public int stringCharacterByteIndex(DynamicObject string, int charIndex) {
+        int singleByteOptimizable(DynamicObject string, int charIndex) {
             return charIndex;
         }
 
         @Specialization(guards = { "!isSingleByteOptimizable(string)", "isFixedWidthEncoding(string)" })
-        public int stringCharacterByteIndexFixedWidthEncoding(DynamicObject string, int charIndex) {
+        int fixedWidthEncoding(DynamicObject string, int charIndex) {
             final Encoding encoding = encoding(string);
             return charIndex * encoding.minLength();
         }
 
-        @Specialization(guards = { "!isSingleByteOptimizable(string)", "!isFixedWidthEncoding(string)" })
-        public int stringCharacterByteIndexMultiByteEncoding(DynamicObject string, int charIndex) {
+        @Specialization(guards = { "!isSingleByteOptimizable(string)", "!isFixedWidthEncoding(string)",
+                "charIndex == 0" })
+        int multiByteZeroIndex(DynamicObject string, int charIndex) {
+            return 0;
+        }
+
+        @Specialization(guards = { "!isSingleByteOptimizable(string)", "!isFixedWidthEncoding(string)",
+                "charIndex == rope(string).characterLength()" })
```

Comment:
You might need or want to guard on the string not having a broken code range. This starts getting strange because MRI doesn't define what to do with broken code ranges, but if you don't do exactly what they they, you end up with spec failures. I say it's strange here because this is code inherited from Rubinius and doesn't really match MRI to begin with. It's something to be aware of at at least.

Reply:
It seems not needed. This case is used in `String#rindex`, and there the `charIndex` is `string.size` and we want to translate that to the `bytesize`.

This example on a broken String:
```ruby
s="B\xe1t"
p s.size
p s.rindex "t"
```
works fine on both MRI and TruffleRuby.